File: trunk/test/R/AT_Test_Katz.R
################################################################################################
# R test script for implemented electron range models
################################################################################################
# Copyright 2006, 2009 Steffen Greilich / the libamtrack team
# 
# This file is part of the AmTrack program (libamtrack.sourceforge.net).
#
# AmTrack is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# AmTrack is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# long with AmTrack (file: copying.txt).
# If not, see <http://www.gnu.org/licenses/>
#
################################################################################################

dev.off

# clear workspace
rm( list = ls() )

# load libAmTrack library
dyn.load("../../Release/libAmTrack.dll")

# load wrapping scripts
source("../../wrapper/R/AmTrack.R")

# necessary library for plotting
library("lattice")

####################### DOSE(DISTANCE) function test #########################################

# radius definitions:
r.m <- 10^seq (-14, -3, length.out=100)

# electron range models definition
ER.model.names <- c("simple test",  "Butts & Katz' (linear)",  "Waligorski's (power-law wmax)",  "Geiss' (power-law E)", "Scholz' (power-law E)", "Edmund' (power-law wmax)","Tabata")
ER.model <- c(2,3,4,5,6,7)

# RDD models definition
RDD.model.names <- c("Simple step test function",  "Katz' point target", "Geiss'", "Site", "Cucinotta", "KatzExtTarget", "CucinottaExtTarget")
RDD.model <- c(1,2,3,4,5,6,7)

# RDD parameters
RDD.parameters <- list(c(1),c(1e-10,1e-10), c(1e-8),c(1e-8,1e-10),c(5e-11,1e-10),c(1e-10,1e-8,1e-10),c(5e-11,1e-8,1e-10))

# data frame setup
df1 <- expand.grid( r.m = r.m, ER.model = ER.model, RDD.model = RDD.model )

df1$ER.model.name	<- as.character(ER.model.names[df1$ER.model])
df1$RDD.model.name	<- as.character(RDD.model.names[df1$RDD.model])

df1$inact.prob	<- numeric(nrow(df1))

df1$r.m.check	<- numeric(nrow(df1))


material.no  <-  c(1)
E.MeV.u      <-  c(100.0)
particle.no  <-  c(1001)

# 1, D0, c, m, 0
GR.parameters <- c(1, 3.0, 1, 2, 0)

for( i in ER.model) {
 ii 			<- df1$ER.model == i
	for( j in RDD.model) {
 	jj 			<- ((df1$RDD.model == j) & (df1$ER.model == i)) 
			df1$inact.prob[jj] <- AT.Katz.inactivation.probability(r.m = df1$r.m[jj], E.MeV.u, particle.no, material.no, rdd.model = j, rdd.parameters = RDD.parameters[[j]], er.model = i, GR.parameters)
		}
}

#df1

# plots...

p1 <- xyplot( inact.prob ~ r.m | ER.model.name , groups = RDD.model.name, ref = TRUE, data=df1, pch = ".", lty = 1, type = "l", xlab = "Distance [m]", ylab = "Inactivation prob.", auto.key = list(title = "Protons in liquid water",points = FALSE, lines = TRUE), scales = list(log = 10))
p2 <- xyplot( inact.prob ~ r.m | RDD.model.name, groups = ER.model.name, ref = TRUE, data=df1, pch = ".", lty = 1, type = "l", xlab = "Distance [m]", ylab = "Inactivation prob.", auto.key = list(title = "Protons in liquid water",points = FALSE, lines = TRUE), scales = list(log = 10))

p1logx <- xyplot( inact.prob ~ r.m | ER.model.name , groups = RDD.model.name, ref = TRUE, data=df1, pch = ".", lty = 1, type = "l", xlab = "Distance [m]", ylab = "Inactivation prob.", auto.key = list(title = "Protons in liquid water",points = FALSE, lines = TRUE), scales = list( x = list(log = 10)))
p2logx <- xyplot( inact.prob ~ r.m | RDD.model.name, groups = ER.model.name, ref = TRUE, data=df1, pch = ".", lty = 1, type = "l", xlab = "Distance [m]", ylab = "Inactivation prob.", auto.key = list(title = "Protons in liquid water",points = FALSE, lines = TRUE), scales = list( x = list(log = 10)))


pdf("Katz.pdf")

p1
p2
p1logx
p2logx

dev.off()

File: trunk/test/C/AT_test_Katz.c
/*
 * @file
 * @brief Dummy file to save dose pattern to output file.
 */

/*
 *    AT_test_GSM.c
 *    ===================
 *
 *    Created on: 2010-06-11
 *    Creator: kongruencja
 *
 *    Copyright 2006, 2010 The libamtrack team
 *
 *    This file is part of the AmTrack program (libamtrack.sourceforge.net).
 *
 *    AmTrack is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    AmTrack is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with AmTrack (file: copying.txt).
 *    If not, see <http://www.gnu.org/licenses/>
 */

#include <stdio.h>
#include <stdlib.h>
#include <malloc.h>
#include <math.h>
#include <stdbool.h>
#include <string.h>

#include "AmTrack.h"
#include "AT_ElectronRange.h"

void AT_IGK_save_survival_curve_to_file( const long  number_of_datapoints,
    const double   max_dose_Gy,
    const double   E_MeV_u,
    const long     particle_no,
    const long     material_no,
    const long     rdd_model,
    const double   rdd_parameter[],
    const long     er_model,
    const double   D0_characteristic_dose_Gy,
    const double   m_number_of_targets,
    const double   kappa,
    const char *   filename){

	FILE * outfile = fopen( filename, "w" );

	double dose_Gy;
	for( dose_Gy = 0 ; dose_Gy < max_dose_Gy ; dose_Gy += max_dose_Gy/number_of_datapoints ){
		double fluence_cm2;
		AT_fluence_cm2(  1,
		    &E_MeV_u,
		    &particle_no,
		    &dose_Gy,
		    material_no,
		    &fluence_cm2);
		double survival = AT_KatzModel_single_field_survival( fluence_cm2, E_MeV_u, particle_no, material_no, rdd_model, rdd_parameter, er_model, D0_characteristic_dose_Gy, m_number_of_targets, kappa );
		fprintf( outfile, "%g %g\n", dose_Gy, survival);
	}

	fclose(outfile);
}


int main( int argc, char * argv[] ){

	char filename1[200] = "sf.dat";

	if( argc != 8){
		printf("Usage: %s nPoints maxDose_Gy E_MeV D0_Gy m a0_m kappa\n", argv[0]);
		printf("\t\texample: %s 50 15.0 10.0 5.0 2.0 1e-8 2000\n", argv[0]);
		exit(EXIT_FAILURE);
	}

	printf("Saving SF to file %s\n", filename1);

	long     number_of_points = atoi(argv[1]);
	double   max_dose_Gy = atof(argv[2]);
	double   E_MeV_u   = atof(argv[3]);

    double   D0_characteristic_dose_Gy = atof(argv[4]);
    double   m_number_of_targets = atof(argv[5]);
    double   a0_m = atof(argv[6]);
    double   kappa = atof(argv[7]);

	long     particle_no = 6012;
	long     material_no = Water_Liquid;
	long     rdd_model   = RDD_KatzExtTarget;
	double   r_min_m     = 1e-10;
	double   D_cutoff_Gy = 1e-40;
	double   rdd_parameter[] = {r_min_m,a0_m,D_cutoff_Gy};
	long     er_model    = ER_Waligorski;

	AT_IGK_save_survival_curve_to_file( number_of_points,
		max_dose_Gy,
	    E_MeV_u,
	    particle_no,
	    material_no,
	    rdd_model,
	    rdd_parameter,
	    er_model,
	    D0_characteristic_dose_Gy,
	    m_number_of_targets,
	    kappa,
	    filename1);

	return EXIT_SUCCESS;
};
